part_two.js

```diff
@@ -1 +1,164 @@
+    87​: { id: ​87​, report_id: ​21​, name: ​'Performance Summary'​, filetype: ​'pdf'​}
+    }, 
+  page: {
+    19​: { id: ​19​, document_id: ​34, body: ​'Lorem ipsum...'​, footnote: ​null​ },
+    21: { id: ​21, document_id: ​8, body: ​'Lorem ipsum...'​, footnote: ​null​ },
+    72​: { id: ​72​, document_id: ​87​, body: ​'Ut aliquet...'​, footnote: ​'Aliquam erat...'​ }, 
+    205​: { id: ​205​, document_id: ​34​, body: ​'Donec a dui et...'​, footnote: ​null​ },
+  },
+}
+
+
+
+/*  Prompt 1 */
+
+function mapReport(store){
+  var result, report, reports, reportId, documents, pages, id;
+
+  reports = store.report;
+  documents = store.document;
+  pages = store.page;
+  result = {};
+
+  // this is assuming that there are checks upstream to ensure clean store val (ie, no dulipcates in this map)
+  for (id in pages){
+    report = reports[documents[pages[id]?.document_id]?.report_id]
+    if (report != undefined){
+      reportId = report.id
+      if (result[reportId]) {
+          ++result[reportId].number_of_pages
+      } else {
+          result[reportId]= {'report_title': report.title, 'report_id': reportId, 'number_of_pages': 1}
+      }
+    }
+  }
+  return result;
+}
+
+/* example function call:
+  mapReport(store)
+ --------------------- */
+
+/* Prompt 2 */
+  
+function mapStore (store){
+  var reports, documents, pages, parentReports;
+
+  reports = store.report;
+  documents = store.document;
+  pages = store.page;
+
+  parentReports = mapObjectsToParentReport( mapObjectsToParentReport(pages, documents, 'document_id'), reports, 'report_id');
+  return (Object.values(parentReports));
+}
+
+function mapObjectsToParentReport (childObject, parentObject, firstParentKeyType){
+  var childId, parent, parentId;
+
+  for (childId in childObject){
+    parentId = childObject[childId][firstParentKeyType]
+    parent = parentObject[parentId]
+    
+    if (parent != undefined){
+      if (parent.children == undefined ) {parent.children = []}
+      if (parent.report_id ){childObject[childId]['report_id'] = parent.report_id}
+      parent.children.push(childObject[childId])
+    }
+  }
+  return parentObject;
+}​
+
+function traverseReportMapBreadthFirst (parentQueue, searchString, searchIn ){
+  var next, search_result, report_ids_list;
+  report_ids_list =[];
+
+  while (parentQueue.length) {
+    next = parentQueue.pop()
+    
+    if (next){
+      search_result = searchObjectForString(next, searchString, searchIn)
+      if (search_result !== undefined){report_ids_list.push(search_result)}
+      else if (next.children !== undefined) {parentQueue = (next.children).concat(parentQueue)}
+    }
+  }
+  return report_ids_list;
+}
+
+function searchObjectForString (nextObject, searchString, searchIn){
+  var property, index;
+
+  for (property of searchIn){
+    index = nextObject[property]?.toLowerCase().indexOf(searchString)
+    if (index !== undefined && index !== -1){
+      return nextObject.report_id || nextObject.id;
+    } 
+  }
+}
```

Comment:
again, don't love this bc it's report-specific; we're assuming that if there's no report id, that must mean we're looking at a report. Decided to use indexOf rather than a regex because it's faster for larger strings.